### PR TITLE
Implemented user friends feature

### LINKED ISSUE #90: Search for and add new friend
**As a** user **I want to** search for and add new friends **so that** I can enhance my friends list.

**Preconditions:**
The registered user is logged in
My Habits page is open

**Acceptance criteria:**

1. A user who doesn’t have friends can search for a new friend by clicking on the button “+” on My habits page:
    a. system redirects him / her to a “Search new friend” tab on My Eco Friends page; 
    b. at All friends tab user can see a message “You don’t have any friends yet”

2. A user who already has friends can search friends by clicking on “See all” button on My habits page and choosing “Search new friend” tab on My Eco friends page.

3. System displays a text “Search for a new friend” in the search field:
    a. text in the input field disappears when user enters the first character and appears again if the input field is empty.

4. User is able to search for a new friend by typing in a search field:
   4.1. One of the following information:
       a. friend’s name;
       b. username.
   4.2. user should type a minimum 1 character and a maximum 30 characters (alphabetic characters, a dot and a space character) and press “Enter”;
      a. system is able to search results from 1 character typed;
      b. system searches results respectively to the order of characters typed in any part of friend’s name;

5. System should display the search results under the “search” field:
  5.1. systems displays results of friends based on the correct combination of typed characters;
  5.2. system displays the “<number> of people found” at the bottom of the search field;
  5.3. system displays no more than 10 results found, to see more user should be able to use infinite scroll;
  5.4. search results are displayed in order of most relevant friends;
  5.5. search results are represented by a list of friends’ cards which contain the following attributes:
    a. avatar 
    b. name
    c. personal rate
    d. city
    e. mutual friends
    f. button “Add friend”
  5.6. by clicking on the “friend’s card” user is redirected to the friend’s page.
  5.7. when system didn’t find any results user sees a message “We didn't find any results matching this search”.
image

6. User is able to filter search results by clicking on the button “Filters” with the following criteria:
    a. City (same city of living)
    b. Friends of friends

7. User is able to apply all filters at the same time.

8. User can add a friend in two ways:
    a. by clicking on the button “add friend” on the friend’s card on the Search results tab;
    b. by clicking on the button “add friend” on the friend’s profile page.

9. When a user clicks on the button “add friend” it is automatically replaced by a button “cancel request” on the friend’s card and on the friend’s profile page until he/she accepts the user's request.

10. User is able to cancel a friendship request by clicking on the button “cancel request”:
    a. it is replaced by a button “Add friend” on the friend’s card and on the friend’s profile page.

**Out of scope:**
Delete friends from my All friends list
Accept or decline friends’ requests

**To be discussed:**
Filter on personal rate diapasons

### LINKED ISSUE #168: Create the service layer for Friends
**Description:**

Implement the service layer responsible for handling all friend-related operations, such as sending friend requests, accepting or declining them, listing friends, and removing friends. This layer will encapsulate the business logic and interact with the repository layer.

**Tasks:**

- [ ] Create FriendService interface
- [ ] Implement FriendServiceImpl class

**Add necessary validations:**

- [ ] Prevent sending a request to oneself
- [ ] Check if users are already friends
- [ ] Prevent duplicate friend requests
- [ ] Implement logic for friendship status transitions (e.g., PENDING → ACCEPTED)
- [ ] Inject and use FriendRepository and UserRepository
- [ ] Write unit tests for all public methods in FriendServiceImpl

**Definition of Done:**

- [ ] All friend-related business logic is implemented in the service layer
- [ ] Each method is covered with unit tests
- [ ] All edge cases and exceptions (e.g., user not found, already friends) are properly handled
- [ ] Code is clean, readable, and follows project conventions

### LINKED ISSUE #169: Create the repository layer for Friends
**Description:**

Implement the repository layer responsible for accessing and managing friendship-related data in the database. This includes friend relationships, friend requests, and search/filter functionality for finding new friends.

**Tasks:**

- [ ] Create FriendRepository interface extending JpaRepository
- [ ] Create custom queries

**Define methods for:**

- [ ] existsFriendRequestBetween(Long senderId, Long receiverId) — checks if a friend request exists between two users.
- [ ] findBySenderIdAndReceiverId(Long senderId, Long receiverId) — finds a friend request between sender and receiver, returns Optional<FriendRequest>.
- [ ] findAllByUserId(Long userId) — returns a list of Friend entities (all friends of the user).
- [ ] searchPotentialFriends(FriendSearchRequestDto request, Pageable pageable) — accepts a DTO with search parameters and pageable, returns a list of FriendCardDto (search results).
- [ ] areFriends(Long userId1, Long userId2) — checks whether the users are friends.
- [ ] deleteFriendRequestBySenderIdAndReceiverId(Long senderId, Long receiverId) — deletes (cancels) a friend request.
- [ ] findUserProfileById(Long id) — returns an Optional<FriendProfileDto> (user profile for friend page).
- [ ] 

**Definition of Done:**

- [ ]  FriendRepository and all necessary query methods implemented
- [ ]  Repository methods are covered with unit or integration tests
- [ ]  Queries are optimized and return correct results
- [ ]  All methods support integration with service layer without redundant logic in service

### LINKED ISSUE #170: Implement GET /api/friends/search endpoint to search potential friends
**Description:**
Create an endpoint that accepts FriendSearchRequestDto query parameters for searching potential friends by name or username, with optional filters by city and friends-of-friends. Supports pagination/infinite scroll with Pageable.
Returns a paginated list of FriendCardDto.

**Request query parameters:**

**FriendSearchRequestDto:**
query (String, required, 1-30 chars)
city (String, optional)
friendsOfFriends (Boolean, optional)
page, size for pagination

**Response:**
List of FriendCardDto (max 10 per page), including avatar, name, personal rate, city, mutual friends, and friend request status.

### LINKED ISSUE #171: Create endpoint POST /api/friends/{id}/add to send a friend request
**Description:**
Allows the current authenticated user to send a friend request to the user with the specified ID.
If the users are already friends or a request has been sent before, return 409 CONFLICT.

**Path variable:**

id — target user ID
Response:

200 OK if request is successfully sent
409 CONFLICT if request already exists or users are already friends

**Access:**
Authenticated users only

### LINKED ISSUE #172: Create endpoint DELETE /api/friends/{id}/cancel to cancel a friend request
Description:
Allows the sender of a friend request to cancel it. The request is removed, and the frontend should update the button state to "Add friend".

Path variable:
id — target user ID

Response:
200 OK on successful cancellation

Access:
Only the sender of the request

### LINKED ISSUE #173: Implement GET /api/friends endpoint to retrieve current user's friends list
**Description:**
Returns all friends of the current authenticated user as a list of FriendCardDto.
Uses findAllByUserId method internally.

**Response:**
List of FriendCardDto

### LINKED ISSUE #174: Implement GET /api/friends/{id} endpoint to retrieve user's profile summary
**Description:**
Returns a short profile of the user identified by id as FriendProfileDto.
Used for displaying friend's profile details when clicking on a friend's card.
Uses findUserProfileById.

**Path variable:**
id — user ID

**Response:**
FriendProfileDto

### LINKED ISSUE #184: Create entities and DTOs for friend management feature

### PR DESCRIPTION
**Pull request by MykytaSirobaba**
Added endpoints, service logic, and database interactions to manage user friends functionality.
Implemented friend requests, acceptance, and removal operations.
Covered with unit tests to ensure reliability and correctness.

### CODE REVIEW

File: dao/src/main/java/greencity/enums/FriendStatus.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package greencity.enums;
+
+public enum FriendStatus {
+    FRIEND,
+    REQUEST,
+    ACCEPTED,
```
Comment (natri0) [inline dao/src/main/java/greencity/enums/FriendStatus.java]: what's the difference between `FRIEND` and `ACCEPTED`? to me it seems like when you accept a request, it goes from `REQUEST` to `FRIEND`.

File: dao/src/main/java/greencity/repository/UserFriendRepository.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+     * Deletes a specific friend request between two users by status.
+     *
+     * @param userId   ID of the user who received the request
+     * @param friendId ID of the user who sent the request
+     * @param status   Friendship status to match (e.g., REQUEST)
+     */
+    @Modifying
+    @Transactional
+    @Query("DELETE FROM UserFriend uf WHERE uf.user.id = :friendId AND uf.friend.id = :userId AND uf.status = :status")
+    void deleteFriendRequest(@Param("userId") Long userId, @Param("friendId") Long friendId, @Param("status") FriendStatus status);
```
Comment (natri0) [inline dao/src/main/java/greencity/repository/UserFriendRepository.java]: does this really need the `FriendStatus` parameter? it's not like we can have multiple UserFriend instances between the same two users with two statuses (e.g. user1&user2 `ACCEPTED` and user1&user2 `BLOCKED` *at the same time*.)

File: dao/src/main/java/greencity/repository/UserFriendRepository.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    FROM UserFriend uf1
+    JOIN UserFriend uf2 ON uf1.friend.id = uf2.friend.id
+    WHERE uf1.user.id = :userId
+      AND uf2.user.id = :targetUserId
+      AND uf1.status = :status
+      AND uf2.status = :status
+""")
+    Long countMutualFriends(@Param("userId") Long userId,
+                            @Param("targetUserId") Long targetUserId,
+                            @Param("status") FriendStatus status);
```
Comment (natri0) [inline dao/src/main/java/greencity/repository/UserFriendRepository.java]: does this really need the `FriendStatus` parameter? i feel like we don't have any use for searching mutual friends that have anything but a `FRIEND` FriendStatus.

File: dao/src/main/java/greencity/repository/UserFriendRepository.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            SELECT ff.friend.id FROM UserFriend f
+            JOIN UserFriend ff ON f.friend.id = ff.user.id
+            WHERE f.user.id = :currentUserId
+              AND ff.status = 'FRIEND'
+              AND ff.friend.id NOT IN (
+                  SELECT fr.friend.id FROM UserFriend fr
+                  WHERE fr.user.id = :currentUserId AND fr.status = 'FRIEND'
+              )
+        ))
+    ORDER BY u.name
```
Comment (natri0) [inline dao/src/main/java/greencity/repository/UserFriendRepository.java]: this would be way more readable if it were separate methods for checking "users in same city", "by mutual friends". also, is there any reason to support *not* excluding the current user?

File: dao/src/main/java/greencity/repository/UserRepo.java [hunk 5]
```diff
@@ -119,25 +119,80 @@ public interface UserRepo extends JpaRepository<User, Long>, JpaSpecificationExe
...
+     *
+     * @param userId   ID of the first user
+     * @param friendId ID of the second user
+     * @return true if users are friends, false otherwise
+     */
+    @Query(nativeQuery = true,
+            value = "SELECT EXISTS(SELECT * FROM users_friends WHERE status = 'FRIEND' AND ("
+                    + "user_id = :userId AND friend_id = :friendId OR "
+                    + "user_id = :friendId AND friend_id = :userId))")
+    boolean isFriend(Long userId, Long friendId);
```
Comment (natri0) [inline dao/src/main/java/greencity/repository/UserRepo.java]: is there any reason for why this is in UserRepo and not UserFriendRepo?

File: dao/src/main/java/greencity/repository/UserRepo.java [hunk 6]
```diff
@@ -119,25 +119,80 @@ public interface UserRepo extends JpaRepository<User, Long>, JpaSpecificationExe
...
+     * This method assumes the request exists in direction from friend to current user.
+     *
+     * @param userId   ID of the current user (receiver of request)
+     * @param friendId ID of the user who sent the friend request
+     */
+    @Modifying
+    @Query(nativeQuery = true,
+            value = "UPDATE users_friends SET status = 'FRIEND' "
+                    + "WHERE user_id = :friendId AND friend_id = :userId")
+    void acceptFriendRequest(Long userId, Long friendId);
```
Comment (natri0) [inline dao/src/main/java/greencity/repository/UserRepo.java]: this should probably be moved into UserFriendRepo.

File: dao/src/main/java/greencity/repository/UserRepo.java [hunk 7]
```diff
@@ -119,25 +119,80 @@ public interface UserRepo extends JpaRepository<User, Long>, JpaSpecificationExe
...
+     * @param userId   ID of one user
+     * @param friendId ID of the other user
+     * @return true if a friend request exists in any direction, false otherwise
+     */
+    @Query(nativeQuery = true,
+            value = "SELECT EXISTS(SELECT * FROM users_friends "
+                    + "WHERE status = 'REQUEST' AND ("
+                    + "(user_id = :userId AND friend_id = :friendId) OR "
+                    + "(user_id = :friendId AND friend_id = :userId)))")
+    boolean isFriendRequested(Long userId, Long friendId);
```
Comment (natri0) [inline dao/src/main/java/greencity/repository/UserRepo.java]: this and the above could be one method like `boolean friendRequestExists(Long userId, Long friendId, FriendStatus status)` in UserFriendRepo.

File: service/src/main/java/greencity/service/FriendServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+        List<User> users = userFriendRepository.searchUsers(currentUserId, query, filterByCity, filterByMutual);
+
+        List<UserCardDto> dtoList = users.stream()
+                .map(user -> {
+                    UserCardDto dto = modelMapper.map(user, UserCardDto.class);
+                    Long mutualCount = userFriendRepository.countMutualFriends(currentUserId, user.getId(), FriendStatus.FRIEND);
+                    dto.setMutualFriendsCount(mutualCount);
+                    dto.setFriend(false);
+                    return dto;
+                })
```
Comment (natri0) [inline service/src/main/java/greencity/service/FriendServiceImpl.java]: i feel like this could be a method in UserRepo that returns UserCardDto that automatically populates mutual friend count.

File: service/src/main/java/greencity/service/FriendServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+     * @throws BadRequestException if request is invalid (already sent, self-request, etc.)
+     * @throws NotFoundException   if either user does not exist
+     */
+    @Override
+    @Transactional
+    public void sendFriendRequest(Long currentUserId, Long friendId) {
+        validateUserAndFriendNotSamePerson(currentUserId, friendId);
+        validateUserAndFriendExistence(currentUserId, friendId);
+        validateFriendNotExists(currentUserId, friendId);
+        validateFriendRequestNotSent(currentUserId, friendId);
```
Comment (natri0) [inline service/src/main/java/greencity/service/FriendServiceImpl.java:70]: those 4 calls are repeated a lot throughout this class; maybe move them into a common `validateFriendRequestParameters` or smth like that?

File: service/src/main/java/greencity/service/FriendServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+     * @throws BadRequestException if already friends or invalid operation
+     * @throws NotFoundException   if request does not exist or users not found
+     */
+    @Override
+    @Transactional
+    public void acceptFriendRequest(Long currentUserId, Long friendId) {
+        validateUserAndFriendNotSamePerson(currentUserId, friendId);
+        validateUserAndFriendExistence(currentUserId, friendId);
+        validateFriendNotExists(currentUserId, friendId);
+        validateFriendRequestSentByFriend(currentUserId, friendId);
```
Comment (natri0) [inline service/src/main/java/greencity/service/FriendServiceImpl.java:88]: like above

File: service/src/main/java/greencity/service/FriendServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+     * @throws BadRequestException if request is invalid or already friends
+     * @throws NotFoundException   if no request found
+     */
+    @Override
+    @Transactional
+    public void cancelFriendRequest(Long currentUserId, Long friendId) {
+        validateUserAndFriendNotSamePerson(currentUserId, friendId);
+        validateUserAndFriendExistence(currentUserId, friendId);
+        validateFriendNotExists(currentUserId, friendId);
+        validateFriendRequestSentByFriend(currentUserId, friendId);
```
Comment (natri0) [inline service/src/main/java/greencity/service/FriendServiceImpl.java:106]: like above

File: service/src/main/java/greencity/service/FriendServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+        Map<Long, Boolean> friendMap = friendIds.stream()
+                .collect(Collectors.toMap(id -> id, id -> true));
+
+        List<UserCardDto> dtos = friends.stream()
+                .map(user -> {
+                    UserCardDto dto = modelMapper.map(user, UserCardDto.class);
+                    dto.setMutualFriendsCount(mutualMap.getOrDefault(user.getId(), 0L));
+                    dto.setFriend(friendMap.getOrDefault(user.getId(), false));
+                    return dto;
+                })
```
Comment (natri0) [inline service/src/main/java/greencity/service/FriendServiceImpl.java]: this is another place where a method in UserRepo that returns a populated UserCardDto would be very helpful.

**Review by natri0 — CHANGES_REQUESTED** [review]

File: dao/src/main/java/greencity/repository/UserFriendRepository.java (hunk 4 above)
Comment (MykytaSirobaba) [reply dao/src/main/java/greencity/repository/UserFriendRepository.java]: In this specific case, breaking the query into separate repository methods for filtering by city and mutual friends doesn’t make much sense. The reason is that the search needs to combine multiple optional filters dynamically, and executing multiple queries would significantly increase the number of database calls, leading to performance issues.

Currently, having a single, parameterized query that applies all filters in one go is more efficient because it executes just one database query regardless of the combination of filters. Splitting the logic would either cause a lot of conditional branching in the service layer or require multiple queries to the database, both of which would complicate the code and hurt performance.

Therefore, keeping the filtering logic together in one query is a better trade-off between maintainability and efficiency in this case.

**Review by Linney34 — APPROVED** [review]

File: service/src/main/java/greencity/service/FriendServiceImpl.java (hunk 9 above)
Comment (MykytaSirobaba) [reply service/src/main/java/greencity/service/FriendServiceImpl.java:70]: I intentionally kept these validation calls separate instead of combining them into a single method like `validateAll()` to preserve semantic clarity. Each validation method checks a specific rule, and calling them explicitly makes the flow of logic more transparent and easier to debug or extend. This also improves readability for other developers by making it clear what exact conditions are being enforced in each case.

File: service/src/main/java/greencity/service/FriendServiceImpl.java (hunk 10 above)
Comment (MykytaSirobaba) [reply service/src/main/java/greencity/service/FriendServiceImpl.java:88]: I intentionally kept these validation calls separate instead of combining them into a single method like `validateAll()` to preserve semantic clarity. Each validation method checks a specific rule, and calling them explicitly makes the flow of logic more transparent and easier to debug or extend. This also improves readability for other developers by making it clear what exact conditions are being enforced in each case.
